### PR TITLE
Expose Terraform Docker provider

### PR DESCRIPTION
**Pull request by swgillespie**
This is very WIP and isn't ready to merge yet.

This PR combines the code currently residing in this package with the [Terraform Docker provider](https://www.terraform.io/docs/providers/docker/index.html), which lets us interact with the Docker daemon to launch containers, download images, create networks, etc. on a host that the running Pulumi program has access to.

This is potentially interesting from a Pulumi perspective in that it 1) sits nicely in a niche currently owned by `docker-compose`, which is the bootstrapping of nontrivial developer environments and 2) it appears to be possible to create a container and request that it run to completion, which could potentially be used as a generalized "provisioner" mechanism for running arbitrary commands on the local machine when certain lifecycle events occur.

In order to be a good citizen with the existing abstractions already in this library, the `docker_image` Terraform resource is projected as `docker.RemoteImage`, since this more clearly identifies the purpose of the resource: it downloads an image from an external repository. Other than that, all other Docker resources are projected verbatim.

I've added two examples in this PR that spin up nginx containers, one in TypeScript and the other in Python.

### CODE REVIEW

**Comment by swgillespie** [comment]
Fixes https://github.com/pulumi/home/issues/386

**Comment by CyrusNajmabadi** [comment]
Sat down with sean.  came up with a few tings to trim down since they don't seem necessary.  Overall though this looks like something very complementary to the docker package.  Once @swgillespie pushes again, it woudl be good if you took a look @lukehoban

**Comment by swgillespie** [comment]
Hi all,

In the last commit I've made a few changes on top of what's already been reviewed, based on my usage of this package:

1. I marked `docker.Container` as `DeleteBeforeReplace`. Docker often consumes single-use resources on the local machine (e.g. ports) and it's annoying to work around these limitations. Marking as DBR makes `docker.Container` significantly less frustrating to use, although it does opt us in to the trouble that DBR provides (https://github.com/pulumi/pulumi/issues/2167).
2. I renamed `networks_advanced` to `networksAdvanced` to fit in better with the camelCase naming convention.

**Comment by swgillespie** [comment]
Would anyone be opposed to merging this today?

**Comment by CyrusNajmabadi** [comment]
Looking now!

File: build/common.mk [hunk 1]
```diff
@@ -36,11 +36,9 @@
 #
 #  - only_build: this target runs build and install targets
 #
-#  - only_test: this target runs the lint and test_all targets
+#  - only_test: this target runs the list and test_all targets
```
Comment (CyrusNajmabadi) [inline build/common.mk]: is that correct?

File: build/common.mk [hunk 2]
```diff
@@ -106,17 +104,52 @@ endif
 PULUMI_BIN          := $(PULUMI_ROOT)/bin
 PULUMI_NODE_MODULES := $(PULUMI_ROOT)/node_modules
 
-.PHONY: default all ensure only_build only_test only_test_fast build lint install test_fast test_all core
+.PHONY: default all ensure only_build only_test build lint install test_fast test_all core
```
Comment (CyrusNajmabadi) [inline build/common.mk]: why did only_test_fast go away?

File: build/common.mk [hunk 3]
```diff
@@ -171,14 +204,13 @@ install::
...
-		(yarn unlink > /dev/null 2>&1 || true) && \
-		yarn link
+	yarn install --offline --production && \
+	(yarn unlink > /dev/null 2>&1 || true) && \
+	yarn link
 endif
 
 only_build:: build install
 only_test:: lint test_all
-only_test_fast:: lint test_fast
```
Comment (CyrusNajmabadi) [inline build/common.mk]: i think these removals aren't actually appropriate.  i was actually just adding them to awsinfra to match pulumi/pulumi

Comment (CyrusNajmabadi) [reply build/common.mk]: @ellismg i can see we're inconsistent across our repos here.  what's teh right pattern we should be following/

File: sdk/nodejs/swarm/config.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
+// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+import * as pulumi from "@pulumi/pulumi";
```
Comment (CyrusNajmabadi) [inline sdk/nodejs/swarm/config.ts]: i thought we didn't want the swarm stuff?

File: sdk/python/pulumi_docker/container.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        __props__['hosts'] = hosts
+
+        __props__['hostname'] = hostname
+
+        if not image:
+            raise TypeError('Missing required property image')
+        __props__['image'] = image
+
+        __props__['labels'] = labels
+
```
Comment (CyrusNajmabadi) [inline sdk/python/pulumi_docker/container.py]: aside.  our generator should prob not have all these newlines.

File: build/common.mk (hunk 1 above)
Comment (swgillespie) [reply build/common.mk]: I'm not sure how this happened, I copied all of this from other repos...

File: build/common.mk (hunk 2 above)
Comment (swgillespie) [reply build/common.mk]: I don't know, this came from another repo

**Comment by ellismg** [comment]
@swgillespie Do you mind applying something like https://github.com/pulumi/pulumi-aws/commit/ef6716cad2f5c7712d80a75f62419ff7a91e95b3 here as well (since we are now adding a provider to this package).  I pushed this everywhere else we had a runtime provider, but didn't realize this work was in flight.

**Comment by swgillespie** [comment]
@ellismg yep, will do.

**Pull request merged**